File: backend/TrackRanker.Api/Models/SessionCompletion.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace TrackRanker.Api.Models;

public sealed class SessionCompletion
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = string.Empty;

    [BsonRepresentation(BsonType.ObjectId)]
    public required string TrainingSessionId { get; set; }

    [BsonDateTimeOptions(Kind = DateTimeKind.Utc)]
    public DateTime CompletedAtUtc { get; set; }

    public int ActualIntensity { get; set; }
    public int PerceivedDifficulty { get; set; }
    public List<RepetitionResult> RepetitionResults { get; set; } = [];
    public SessionReflection Reflection { get; set; } = new();

    [BsonDateTimeOptions(Kind = DateTimeKind.Utc)]
    public DateTime CreatedAtUtc { get; set; }

    [BsonDateTimeOptions(Kind = DateTimeKind.Utc)]
    public DateTime UpdatedAtUtc { get; set; }
}

File: frontend/src/types/sessionCompletion.ts
export interface RepetitionResult {
  setNumber: number;
  repetitionNumber: number;
  distanceMetres: number;
  timeSeconds: number;
  notes: string | null;
}

export interface SessionReflection {
  wentWell: string | null;
  improved: string | null;
  wasDifficult: string | null;
  nextFocus: string | null;
  coachFeedback: string | null;
  confidenceBefore: number | null;
  confidenceAfter: number | null;
}

export interface SessionCompletionInput {
  actualIntensity: number;
  perceivedDifficulty: number;
  repetitionResults: RepetitionResult[];
  reflection: SessionReflection;
}

export type CreateSessionCompletionRequest = SessionCompletionInput;
export type UpdateSessionCompletionRequest = SessionCompletionInput;

export interface SessionCompletion extends SessionCompletionInput {
  id: string;
  trainingSessionId: string;
  completedAtUtc: string;
  createdAtUtc: string;
  updatedAtUtc: string;
}
